umad/src/main/java/com/pkukielka/AccessMonitorRewriter.java

```diff
@@ -79,15 +90,105 @@ public static void logUnsafeMethodCalls(String methodName, String ifCalledFrom,
         }
     }
 
-    public AccessMonitorRewriter(Config config){
+    public static Set<String> getMethodsMarkedAsSafe() {
+        return methodsMarkedAsSafe;
+    }
+
+    public AccessMonitorRewriter(Config config) {
         super(config);
+        if (isEnabled()) {
+            String msg = "Analyzing method access with shouldThrowExceptions=%s and shouldPrintStackTrace=%s";
+            System.out.println(String.format(msg, conf.shouldThrowExceptions, conf.shouldPrintStackTrace));
+
+            for (Config safeIndicator : config.getConfigList("safeIndicators")) {
+                final String clazz = safeIndicator.getString("clazz");
+                final String jvmType = "L" + clazz.replace('.', '/') + ";";
+                safeFieldTypes.add(jvmType);
+                safeMethodTypes.add("()" + jvmType); // Getter of thread local
+                for (Object methodTpe : safeIndicator.getAnyRefList("methods"))
+                    safeMethods.add(clazz + "." + methodTpe);
+            }
+
+            synchronizeIndicators.addAll(config.getStringList("synchronizeIndicators"));
+        }
     }
 
+
+    private final String callLogUnsafeMethodCalls = "com.pkukielka.AccessMonitorRewriter.logUnsafeMethodCalls" +
+            "(\"%s\", \"%s\", System.identityHashCode(this));";
+
     @Override
-    protected void editMethod(final CtMethod editableMethod, String ifCalledFrom) throws CannotCompileException {
+    protected void editMethod(CtMethod editableMethod, String ifCalledFrom, String dottedName) throws CannotCompileException {
         String methodName = editableMethod.getLongName();
-        editableMethod.insertBefore(
-                String.format("com.pkukielka.AccessMonitorRewriter.logUnsafeMethodCalls(\"%s\", \"%s\", System.identityHashCode(this));",
-                        methodName, ifCalledFrom));
+        editableMethod.insertBefore(String.format(callLogUnsafeMethodCalls, methodName, ifCalledFrom));
+    }
+
+    @Override
+    public void applyOnMethod(CtMethod editableMethod, String dottedName) throws CannotCompileException {
+        if (isSafe(editableMethod, dottedName)) {
+            if (conf.shouldStoreSafeMethod) methodsMarkedAsSafe.add(editableMethod.getLongName());
+        } else super.applyOnMethod(editableMethod, dottedName);
+
+    }
+
+    private static Set<Integer> methodOpcodes = new HashSet<Integer>(Arrays.asList(
+            Opcode.INVOKEVIRTUAL,
+            Opcode.INVOKESPECIAL,
+            Opcode.INVOKEINTERFACE,
+            Opcode.INVOKESTATIC
+    ));
+
+    private boolean isThreadLocalGetOrSet(ConstPool pool, int index) {
```

Comment:
I see nothing specific to `ThreadLocal` inside that method. Consider renaming it?